File: csharp/Patterns/Decorator.cs
using System;

namespace DesignPatterns.Patterns;

public static class Decorator
{
    private interface IIceCream
    {
        decimal Cost();
        string Description();
    }

    private class Vanilla : IIceCream
    {
        public decimal Cost() => 2.0m;
        public string Description() => "vanilla";
    }

    private abstract class ToppingDecorator : IIceCream
    {
        protected ToppingDecorator(IIceCream baseIceCream)
        {
            BaseIceCream = baseIceCream;
        }

        protected IIceCream BaseIceCream { get; }
        public abstract decimal Cost();
        public abstract string Description();
    }

    private class Sprinkles : ToppingDecorator
    {
        public Sprinkles(IIceCream baseIceCream) : base(baseIceCream) { }

        public override decimal Cost() => BaseIceCream.Cost() + 0.5m;
        public override string Description() => BaseIceCream.Description() + ", sprinkles";
    }

    private class Cherry : ToppingDecorator
    {
        public Cherry(IIceCream baseIceCream) : base(baseIceCream) { }

        public override decimal Cost() => BaseIceCream.Cost() + 0.75m;
        public override string Description() => BaseIceCream.Description() + ", cherry";
    }

    public static void Run()
    {
        Console.WriteLine("\nDecorator");
        IIceCream iceCream = new Vanilla();
        IIceCream fancy = new Cherry(new Sprinkles(iceCream));
        Console.WriteLine($"{fancy.Description()} costs ${fancy.Cost():0.00}");
    }
}

File: python/patterns/decorator.py
from abc import ABC, abstractmethod


class IceCream(ABC):
    @abstractmethod
    def cost(self) -> float:
        raise NotImplementedError

    @abstractmethod
    def description(self) -> str:
        raise NotImplementedError


class Vanilla(IceCream):
    def cost(self) -> float:
        return 2.0

    def description(self) -> str:
        return "vanilla"


class ToppingDecorator(IceCream):
    def __init__(self, base: IceCream) -> None:
        self.base = base


class Sprinkles(ToppingDecorator):
    def cost(self) -> float:
        return self.base.cost() + 0.5

    def description(self) -> str:
        return self.base.description() + ", sprinkles"


class Cherry(ToppingDecorator):
    def cost(self) -> float:
        return self.base.cost() + 0.75

    def description(self) -> str:
        return self.base.description() + ", cherry"


def run() -> None:
    print("\nDecorator")
    ice_cream = Vanilla()
    fancy = Cherry(Sprinkles(ice_cream))
    print(f"{fancy.description()} costs ${fancy.cost():.2f}")
